Fix fast logging

rowleya opened the pull request:
Attempt to fix #62 by reducing the rate of uploads.

rowleya committed:
Attempt to fix issues with logging that can freeze jobs

rowleya committed:
Style

rowleya committed:
Try to avoid having logs being updated in two places!

rowleya committed:
Mark when log update in progress in case of debug issues

rowleya committed:
Merge branch 'master' into fix_fast_logging

rowleya committed:
Remove import (not used anyway)

rowleya committed:
Catch a throwable to make sure *everything* gets caught

rowleya committed:
Stop logs becoming too big

rowleya committed:
Catch this upstream

rowleya committed:
Add the docs

rowleya committed:
Remove docs for params that have been removed

rowleya committed:
Flush logs during run and put warning in the output too

rowleya committed:
Remove unused class

rowleya committed:
Move files to where they are used

rowleya committed:
Use the right type, and get the working directory from the parameters

rowleya committed:
Add the log file to the outputs

rowleya committed:
Use the txt ending to simplify the display

rowleya committed:
Handle output files in a better way

rowleya committed:
Fix style

dkfellows commented on RemoteSpiNNaker/RemoteSpiNNakerModel/src/main/java/uk/ac/manchester/cs/spinnaker/job/JobManagerInterface.java:
```diff
@@ -246,56 +248,39 @@ void addProvenance(@PathParam("id") int id,
      */
     @POST
-    @Path("{projectId}/{id}/finished")
+    @Path("{id}/finished")
```
I like this change very much.

dkfellows commented on RemoteSpiNNaker/RemoteSpiNNakerJobProcessManager/src/main/java/uk/ac/manchester/cs/spinnaker/jobprocessmanager/JobProcessManager.java:
```diff
@@ -100,57 +101,107 @@
-                    sendLog();
+        public void append(final String logMsg) throws IOException {
+            logFileWriter.append(logMsg);
+            logFileWriter.flush();
```
Is this call really necessary? It's potentially expensive.

dkfellows replied:
In general, you only need to `flush` when you're going to hand off to another process without closing (or if you're mad enough to have several streams looking at the same file at once from one process when at least one of those is a writer).

rowleya replied:
This was because the log is no longer written anywhere but here, so flushing here makes it easier to debug when looking at the output on the machine.

dkfellows commented on RemoteSpiNNaker/RemoteSpiNNakerWeb/src/main/java/uk/ac/manchester/cs/spinnaker/jobmanager/JobManager.java:
```diff
@@ -202,6 +202,35 @@
+     *
+     * Use this to synchronise on log updates actually happening.
+     */
+    private final ActionListener logTimerAction = new ActionListener() {
```
This could be a lambda as you're using the Java 8 language profile:
```java
private final ActionListener logTimerAction = (event -> sendLogs());
```
(Outer parentheses may be optional. The style will know.)

dkfellows commented on RemoteSpiNNaker/RemoteSpiNNakerWeb/src/main/java/uk/ac/manchester/cs/spinnaker/output/OutputManagerImpl.java:
```diff
@@ -243,44 +241,33 @@ private File getProjectDirectory(final String projectId) {
-                if (outputPath.startsWith("/")) {
-                    outputPath = outputPath.substring(1);
-                }
+            if (!output.startsWith(baseDirectory)) {
```
You also need a check if the name contains `..` as a component.

And you ought to do the name sanity checks before acquiring the lock; they don't (and shouldn't) depend on the state of the named entities.

dkfellows reviewed (commented):
Doesn't look too bad. Few minor quibbles.

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Add argument validation to the output manager

Failure to validate arguments from a potentially non-controlled source
is one of the biggest sources of security bugs there is.

dkfellows committed:
checkstyle

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Make more local variables final

It's not my style choice, but if we're doing it we should do it.

dkfellows committed:
Style

dkfellows committed:
Style

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging

dkfellows committed:
Merge branch 'master' into fix_fast_logging
